lib/worker.js

```diff
@@ -0,0 +1,211 @@
+    if (typeof ErrorConstructor !== 'function')
+      ErrorConstructor = Error;
+    var error = new ErrorConstructor(payload.message);
+    error.stack = payload.stack;
+    util._extend(error, payload.additionalProperties);
+    workerObject.emit('error', error);
+  };
+
+  workerContext._onexit = function(exitCode) {
+    workerObject._workerContext = null;
+    workerObject.emit('exit', exitCode);
+  };
+
+  workerContext._onmessage = function(payload, messageType) {
+    payload = JSONParse(payload);
+    switch (messageType) {
+      case WorkerContextBinding.USER:
+        return workerObject.emit('message', payload);
+      case WorkerContextBinding.INTERNAL:
+        assert.fail('unreachable');
+      case WorkerContextBinding.EXCEPTION:
+        return onerror(payload);
+      default:
+        assert.fail('unreachable');
+    }
+  };
+};
+
+Worker.prototype._checkAlive = function() {
+  if (!this._workerContext)
+    throw new RangeError('this worker has been terminated');
+};
+
+Worker.prototype._doPostMessage = function(payload, transferList, type) {
+  this._checkAlive();
+
+  if (transferList == null)
+    transferList = EMPTY_ARRAY;
+  else if (!Array.isArray(transferList))
+    throw new Error('transferList must be an array');
+
+  this._workerContext.postMessage(JSONStringify(payload), transferList, type);
+};
+
+Worker.prototype.postMessage = function(payload, callback) {
+  this._doPostMessage(payload, null, WorkerContextBinding.USER);
+};
+
+Worker.prototype.terminate = function(callback) {
+  this._checkAlive();
+  var context = this._workerContext;
+  this._workerContext = null;
+  if (typeof callback === 'function') {
+    this.once('exit', function(exitCode) {
+      callback(null, exitCode);
+    });
+  }
+  context.terminate();
+};
+
+Worker.prototype.ref = function() {
+  this._checkAlive();
+  this._workerContext.ref();
+};
+
+Worker.prototype.unref = function() {
+  this._checkAlive();
+  this._workerContext.unref();
+};
+
+if (process.isWorkerInstance) {
+  const postMessage = function(payload, transferList, type) {
+    if (transferList == null)
+      transferList = EMPTY_ARRAY;
+    else if (!Array.isArray(transferList))
+      throw new Error('transferList must be an array');
+
+    WorkerContextBinding.workerWrapper._postMessage(JSONStringify(payload),
+                                                    transferList,
+                                                    type);
+  };
+  const workerFatalError = function(er) {
+    var defaultStack = null;
+    var defaultMessage = '[to string conversion failed]';
+    var defaultBuiltinType = 'Error';
+
+    var message = defaultMessage;
+    var builtinType = defaultBuiltinType;
+    var stack = defaultStack;
+    var additionalProperties = {};
+
+    if (er instanceof Error) {
+      try {
+        builtinType = er.name;
+        if (typeof builtinType !== 'string')
+          builtinType = defaultBuiltinType;
+      } catch (ignore) {}
```

Comment:
A real `.name` is always a string, that is being checked too. Generally you get string out of anything but there is still problematic values like e.g. `Object.create(null)` which throws when trying to convert to string.